tools/block-generator/generator/generate.go

```diff
@@ -875,35 +1120,206 @@ func (g *generator) txnForRound(round uint64) uint64 {
 	return g.config.TxnPerBlock
 }
 
-// finishRound tells the generator it can apply any pending state.
-func (g *generator) finishRound(txnCount uint64) {
-	g.txnCounter += txnCount
+// startRound updates the generator's txnCounter based on the latest block header.
+// It is assumed that g.round has already been incremented in finishRound()
+func (g *generator) startRound() error {
```

Comment:
Are we unable to manually track the txn counter? There should be a known number of inner transactions. Does it match how many the ledger things there are?

Reply:
This is a great question and deserves a good answer. I'll investigate.

Reply:
After some investigation, I believe it would be difficult to achieve this without some significant downsides. This involves opening up `BlockEvaluator.Eval()` to get the `TxnCounter` directly from the ledger. There is -in fact- a method [BlockEvaluator.TestingTxnCounter()](https://github.com/algorand/go-algorand/blob/8ba0dc67f8a57ba07f510e5680b0968e0512d3c3/ledger/eval/eval.go#L1256) which is meant for this purpose. However, accessing it means interacting with the `BlockEvaluator` that created during the ledger's call to `eval.Eval()`. One relatively non-invasive way I see of achieving this is to modify `interface EvalTracer`'s `AfterBlock()` to take in a new parameter `newTxnCounter` that will let us trace this value and then use it in the new method that I'm introducing [ledgerAddBlock()](https://github.com/tzaffi/go-algorand/blob/291d30fe472810a195c2ccdd12be70ebd213b6b7/tools/block-generator/generator/generator_ledger.go#L56-L88). This feels like overkill to me though. Notice that the same method does introduce an improvement: we calculate the number of transactions in the block by looking at all the actual transactions including inners that were evaluated.

Reply:
Reading a bit more carefully the original comment, it's arguable that the follow up PR _DOES_ address the concern as it:
1. gets the theoretical transaction count using `effects`
2. gets the actual transaction count from `ledgerAddBlock()`
3. compares them and errors if their unequal

see:
https://github.com/tzaffi/go-algorand/blob/f631aec855695e0e984008bcafd1041a46fa0a15/tools/block-generator/generator/generate.go#L382-L388